docs/source/app-dev/app-arch.rst

```diff
@@ -133,22 +133,132 @@ Command deduplication
-- If you are not sure whether a command was submitted successfully, just resubmit it. If the new command was submitted within the deduplication time window, the duplicate submission will safely be ignored. If the deduplication time window has passed, you can assume the command was lost or rejected and a new submission is justified.
+No deduplication needed
+=======================
 
+Some applications do not need command deduplication, say because the Daml models already prevent duplication.
 
-For more details on command deduplication, see the :ref:`Ledger API Services <command-submission-service-deduplication>` documentation.
+* Set a fresh command ID for each submission, e.g., a random UUID.
+* Omit the submission ID, submission rank, and deduplication period.
+  The participant or ledger sets them at their discretion.
+  Under unlucky circumstances, such a submission may result in a ``NOT_FOUND`` or ``OUT_OF_RANGE`` error due to inappropriate choices for the submission rank and deduplication period.
+  If so, the application should resubmit the command.
+
+If the application receives a ``DEADLINE_EXCEEDED`` response or no response at all, then it must not make any assumptions about the command being applied or not until a completion event for the change ID appears on the completion stream.
+Yet, if the submission is actually lost, no such completion event will ever appear.
+To obtain finality, the application may resubmit the command and rely on its own mechanisms (e.g. in the Daml model) to avoid prevent duplication.
+
+
+Bounded network delays and clock skews
+======================================
+
+In this scenario, the application wants to have its intended ledger changes applied at most once and it wants finality about whether they were applied.
+We assume bounds on the following:
+
+* The clock skews between application, participant and ledger.
+  
+* The message delay on the network from the application to the participant and the latency for processing a submission.
+
+* The downtime or failover time of the application system, the particpant, and the ledger.
+
+Commands may be duplicated and finality violated if the actual skews and delays exceed the bounds.
+Finality may not be acheiveable if unhealthy conditions persist for longer than the :ref:`max deduplication time <com.daml.ledger.api.v1.LedgerConfiguration.max_deduplication_time>`.
+
+We further assume that the application persists its intended ledger changes and whether it has decided to not retry a change.
+It also maintains a watermark of the changes that have either been applied or rejected and will not be retried and the time when it has advanced the watermark.
```

Comment:
I cannot easily associate a mental model with this sentence. Would it be possible to give a more concrete example of the kind of watermark that you'd persist? For example, by concretizing the example with the scenario of ingesting from Kafka?

Reply:
That's a good suggestion for a tutorial example. We'll need some hands-on examples with working code.